### PR TITLE
SI-6751 macroContext factory is now public

### PR DESCRIPTION
**Pull request by xeno-by**
In order to make macro contexts customizable by the embedding environment
such as an IDE.

The motivation for this commit was a request to expose an IDE-friendly
debug log API for macros. With macroContext being overridable (see the
instructions in the comments), we can recommend people to use c.info
and then the IDE could override it and present it conveniently.

### CODE REVIEW

**Comment by xeno-by** [comment]
review @retronym @dragos

**Comment by retronym** [comment]
lgtm, although @dragos will need to confirm that he can hook in to this.

**Comment by xeno-by** [comment]
PLS REBUILD jenkins job pr-scala-testsuite-linux-opt

**Comment by adriaanm** [comment]
ping @dragos

File: src/compiler/scala/tools/nsc/typechecker/Macros.scala [hunk 1]
```diff
@@ -550,7 +550,17 @@ trait Macros extends scala.tools.reflect.FastTrack with Traces {
...
+   *
+   *  `typer` is the typer at the callsite of a macro, `prefixTree` is the prefix of the macro application
+   *  (see the documentation for Context.prefix for more info), `expandeeTree` is the macro application itself.
+   *
+   *  This method is designed to be overridable, e.g. from an IDE, which would redirect
+   *  methods like c.echo or c.info to a custom GUI. To override this method, extend `Global` and provide
+   *  an overriden `analyzer` val with an overriden `newTyper` method, which produces a typer with an overriden
+   *  `macroContext` method.
+   */
+  def macroContext(typer: Typer, prefixTree: Tree, expandeeTree: Tree): MacroContext =
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:563]: I might be missing something, but why isn't the usual reporter enough? We already provide our own reporter, so I'd prefer to use that one if possible.

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:563]: So these methods are completely compatible with the IDE? https://github.com/scala/scala/blob/0acb8a30c379f268e8a3e1340504530493a1a1dc/src/compiler/scala/reflect/macros/runtime/FrontEnds.scala#L1

Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:563]: Also when we implement caching, I guess, the IDE might want to plug its own implementation

Comment (dragos) [inline src/compiler/scala/tools/nsc/typechecker/Macros.scala:563]: I think so. We are using this class: https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/interactive/InteractiveReporter.scala As long as messages get there, we can display them. Right now they would appear as error markers in the editor, but only if the position is correctly set. Everything is also spit out in the console, if the presentation debug flag is set.

**Comment by xeno-by** [comment]
Okay, from what I understand Iulian doesn't need this patch right now. Yet I suggest we still merge it, if only for the comment which elaborates on how the macro engine can be customized.

**Comment by gkossakowski** [comment]
What about being able to distinguish between vanilla compile error and errors coming from macros? Can we do that easily?

**Comment by xeno-by** [comment]
@gkossakowski I don't think that it's possible without overriding macro context:

```
def error(pos: Position, msg: String): Unit = callsiteTyper.context.error(pos, msg)
```

**Comment by gkossakowski** [comment]
@dragos wouldn't be useful to distinguish between those to in IDE?

**Comment by xeno-by** [comment]
Ping

**Comment by xeno-by** [comment]
Hey

**Comment by xeno-by** [comment]
Please feel free to reopen if the change proves useful at some point.

**Pull request closed without merging**